xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java

```diff
@@ -0,0 +1,343 @@
+            } catch (Throwable e) {
+                logger.warn("Unexpected exception is raised while building a route using {}" +
+                            ". Try watching after {} seconds..", entry, BACKOFF_SECONDS, e);
+                return false;
+            }
+        }
+
+        setNewSnapshot(cache, ResourceType.ROUTE,
+                       CentralDogmaSnapshotResources.create(routesBuilder.build(), revision));
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void setNewSnapshot(SimpleCache<String> cache, ResourceType resourceType,
+                                SnapshotResources<?> resources) {
+        snapshotLock.lock();
+        try {
+            SnapshotResources<Cluster> clusters;
+            SnapshotResources<ClusterLoadAssignment> endpoints;
+            SnapshotResources<Listener> listeners;
+            SnapshotResources<RouteConfiguration> routes;
+            SnapshotResources<Secret> secrets;
+            final Snapshot previousSnapshot = cache.getSnapshot(DEFAULT_GROUP);
+            if (previousSnapshot == null) {
+                final SnapshotResources<?> emptyResources =
+                        SnapshotResources.create(ImmutableList.of(), "empty_resources");
+                clusters = (SnapshotResources<Cluster>) emptyResources;
+                endpoints = (SnapshotResources<ClusterLoadAssignment>) emptyResources;
+                listeners = (SnapshotResources<Listener>) emptyResources;
+                routes = (SnapshotResources<RouteConfiguration>) emptyResources;
+                secrets = (SnapshotResources<Secret>) emptyResources;
+            } else {
+                clusters = previousSnapshot.clusters();
+                endpoints = previousSnapshot.endpoints();
+                listeners = previousSnapshot.listeners();
+                routes = previousSnapshot.routes();
+                secrets = previousSnapshot.secrets();
+            }
+            switch (resourceType) {
+                case CLUSTER:
+                    clusters = (SnapshotResources<Cluster>) resources;
+                    break;
+                case ENDPOINT:
+                    endpoints = (SnapshotResources<ClusterLoadAssignment>) resources;
+                    break;
+                case LISTENER:
+                    listeners = (SnapshotResources<Listener>) resources;
+                    break;
+                case ROUTE:
+                    routes = (SnapshotResources<RouteConfiguration>) resources;
+                    break;
+                case SECRET:
+                    secrets = (SnapshotResources<Secret>) resources;
+                    break;
+                default:
+                    // Should never reach here.
+                    throw new Error();
+            }
+
+            cache.setSnapshot(DEFAULT_GROUP,
+                              new CentralDogmaSnapShot(clusters, endpoints, listeners, routes, secrets));
```

Comment:
Question) Do we also want to validate the snapshot before setting the new snapshot? Or do we allow inconsistent snapshots here and do validation when the write occurs in the dogma repo? (Or do we do no validation at all)

i.e.
```java
CentralDogmaSnapShot snapshot = ...
snapshot.ensureConsistent();
```

Reply:
Oops, I think I don't get what you meant. 😓 What's the inconsistency you mean in this context?

Reply:
In short, if a cluster references an endpoint that doesn't exist do we allow that update?

Reply:
That's something I missed. 😅 
Let me implement the validation logic in the follow-up PR with the Central Dogma cache.